### ISSUE TITLE
Update ESP8266_RTOS_SDK v1.4.0 to ESP8266_RTOS_SDK v3.x.x

### ISSUE THREAD

**Issue by rajeshpachaikani** [platformio/platform-espressif8266#219]
The version being used is awfully old and no documentation available for it on Espressif's website.

**Comment by BluRaf** [comment]
This problem still exists, I had to migrate project to standard Espressif RTOS SDK because of outdated version used in PlatformIO.

**Comment by Seramis** [comment]
Yup, i write here to give support for updating the SDK for ESP8266.

**Comment by Seramis** [comment]
#85 
#125 
#192 

So, any plans?

**Comment by pamafe1976** [comment]
I am also interested in this update

**Comment by yooshky** [comment]
ESP8266_RTOS_SDK v1.4.0 is unuseable, PIO for me too.

**Comment by shreekant-sureliya** [comment]
is there any update...?

**Comment by utkarshsethi** [comment]
@ivankravets @valeros please help with this

**Comment by randaljunior** [comment]
Is It too hard to make this jump?
no one on PlatformIO cares?

**Comment by mtlcrusher** [comment]
I would like to help, however, I don't know where is the repos that supposed to be edited?

**Comment by koluckirafal** [comment]
> I would like to help, however, I don't know where is the repos that supposed to be edited?

https://github.com/platformio/platform-espressif8266/

1. [platform.json](https://github.com/platformio/platform-espressif8266/blob/develop/platform.json#L58-L63) - here you need to bump version, currently release v3.4: https://github.com/espressif/ESP8266_RTOS_SDK/releases/tag/v3.4
2. [builder/frameworks/esp8266-rtos-sdk.py](https://github.com/platformio/platform-espressif8266/blob/develop/builder/frameworks/esp8266-rtos-sdk.py) - builder script needs to be updated for the RTOS SDK 3.x, which had some breaking changes to make it closer to the ESP-IDF framework. You might want to check how is ESP-IDF support implemented here: https://github.com/platformio/platform-espressif32/blob/develop/builder/frameworks/espidf.py

**Comment by seancale** [comment]
Still looks like it's trying to use the same ancient SDK.

My platform.json has this, even for the development version:
```   
 "framework-esp8266-rtos-sdk": {
      "type": "framework",
      "optional": true,
      "owner": "platformio",
      "version": ">=1.5.0-beta"
    },```

I'm running PlatformIO Core version 6.0.2 on VS Code 1.67.2. Device is a generic ESP8285.

**Comment by bwjohns4** [comment]
Any update on this?

**Comment by uGiFarukh** [comment]
Any update?

**Comment by r315** [comment]
Does platform.io will ever support ESP8266_RTOS_SDK v3.x ?

**Comment by hernot** [comment]
There is a pending pull-request #284 from @freedib which would benefit from some more up-votes and commets from eagerly awaiting users.

**Comment by jeecrypt** [comment]
I would like to return to PlatformIO, but the old version of RTOS SDK does not allow me to do this. I'm really looking forward to the update. I want stability.

**Comment by bebek** [comment]
Is there any commitment to do the SDK upgrade or are we facing a dead end?
We cannot use anymore PIO for the esp8266 with RTOS.

**Comment by g0eki** [comment]
I am also interested and very grateful.

**Comment by hernot** [comment]
Lately while testing the pull request #284  by @freedib I  came across an aspect/issue which might end up as a showstopper on upgrading esp8266 RTOS SDK to  IDF-SDK style (compatible) 3.x sdk . The new SDK seems like IDF-SDK to support only  esp8266 chips and boards with >= 1M flash. When running
```
pio run -t menuconfig
```
one can select between various flash sizes and how they are partitioned with OTA or without eg 
```
1M -> 512/512
2M-> 1024/1024
...
```
There is no option for boards and chips with less than 1M flash eg. the [Sparkfun Thing](https://www.sparkfun.com/sparkfun-esp8266-thing.html).

Therefore i do suggest to consider the ESP8266 rtos sdk < 3.x and the ESP rtos sdk as different not relate sdk's. For example name the 3.x "ESP8266-IDF-SDK" ( pull request #284  ) and keep the name "ESP8266-RTOS-SDK" for version 1.4 or 1.5beta. Would at least be more transparent than implicitly enforcing versions < 3.x if the flash on the platform/board/chip is less than 1M and the 3.x in any other case.

### Rational: 
As long as the binary remains small <~ 380kB the latest 3.x esp8266 rtos sdk will generate a binary which will work like a charm. In my case as long my code just made the onboard led blink and drive my 48 led dot start led strip everything was fine. As soon as i added network and mqtt component to my project the compiled firmware binary exceeded the 512kb flash of my esp8266 based board with > 600kb by far. Uploading this oversized binary failed with a checksum error. When Implementing blinking led, led-strip, network, mqtt and a poor mans touch sensitive pin using the outdated esp8266 rtos SDK 1.4, the resulting binary has less than 330kb and perfectly fits in the program flash. No checksum error. 
### Other Considerations
 The encountered issues might rather bee an issue of configuration tool then the sdk itself. Than the latest (3.x) esp8266 rtos sdk would be capable of generating multi-component (led, ledstrip, network, mqtt, etc) for devices/boards/cips with less than 1M flash memory, while the config tool lacks appropriate entries and parameters for flash memory partitioning and program segment structure for < 1M. In that case filing a feature request or even bug report on the  [esp8266 rtos sdk repo](https://github.com/espressif/ESP8266_RTOS_SDK) and postponing upgrading the sdk version in pio until resolved would be the appropriate approach. Unless Espressive clearly states that they do not further support platforms/boards wich do not extend 512kb onchip flash to at least 1M by adding >= 512 SPI flash.

**Comment by hernot** [comment]
by the way issue #125 seems to request the same.

**Comment by freedib** [comment]
Interesting issue. When I implemented IDF style RTOS, I removed support for 1.5 to avoid naming confusion.
I don't understand why Espressif did not name the new SDK: ESP8266-IDF.
It could easily bring it back. 
I suggest to keep the framework name "esp8266-rtos-sdk" for RTOS 3.4 to be coherent with Espressif  and use something like "esp8266-rtos15-sdk" or "esp8266-rtos-legacy-sdk" for RTOS 1.5

**Comment by freedib** [comment]
You could create a partition.csv file for 512k to get a warning at compile time, but I suspect that components libraries in IDF style are more complex and bigger than legacy ones, and as experimented, except basic programs it won't fit in 512k.

**Comment by hernot** [comment]
> Interesting issue. When I implemented IDF style RTOS, I removed support for 1.5 to avoid naming confusion. I don't understand why Espressif did not name the new SDK: ESP8266-IDF. It could easily bring it back.


A wild blind guess. The esp8266 3.x rtos sdk is meant as intermediate release until native esp8266 ESP-IDF support is finished and thus has nothing in common with the latter apart from API compatibility. 

 
> I suggest to keep the framework name "esp8266-rtos-sdk" for RTOS 3.4 to be coherent with Espressif and use something like "esp8266-rtos15-sdk" or "esp8266-rtos-legacy-sdk" for RTOS 1.5


I fully do support your suggestion.


> You could create a partition.csv file for 512k to get a warning at compile time, but I suspect that components libraries in IDF style are more complex and bigger than legacy ones, and as experimented, except basic programs it won't fit in 512k.


Just a quick question is there any **pio** or **Espressif** tool to extract `partition.csv` from `eagle.app.v6.ld` linker script or what is the safest approach to do it manually? Which pitfalls and human fails are to be considered when doing manually?

**Comment by freedib** [comment]
I don't think such a tool exists and I never modified a partition table but according to https://docs.espressif.com/projects/esp8266-rtos-sdk/en/latest/api-guides/partition-tables.html I would try the following partition.csv:

Name,   Type, SubType, Offset,  Size
nvs,      data, nvs,     0x9000,  0x3000
factory,  app,  factory, 0xC000, 0x74000

ESP8266 is almost discontinued and replaced by ESP32-C3. The platform upgrade I proposed 2 years ago made sense at that time, but now, I suspect Platformio people won't put energy in supporting  a new platform for a product which is in end of life. So I guess RTOS 1.5 will stay the standard on Platformio.

**Comment by hernot** [comment]
> I don't think such a tool exists and I never modified a partition table but according to https://docs.espressif.com/projects/esp8266-rtos-sdk/en/latest/api-guides/partition-tables.html I would try the following partition.csv:
> 
> Name, Type, SubType, Offset, Size nvs, data, nvs, 0x9000, 0x3000 factory, app, factory, 0xC000, 0x74000
> 
> ESP8266 is almost discontinued and replaced by ESP32-C3. The platform upgrade I proposed 2 years ago made sense at that time, but now, I suspect Platformio people won't put energy in supporting a new platform for a product which is in end of life. So I guess RTOS 1.5 will stay the standard on Platformio.

Good argument. So how much efforts should be put into a `partition.csv` for low flash ( <1M ) chips and board anyway? Would be anything beyond a NOTE/WARNING in the README.md in your esp8266 rtos sdk 3.x Repo and related platform  repos worth the efforts at all?

**Comment by freedib** [comment]
I have no Sparkfun Thing to test but the proposed partitions.csv seems to work with standard idf.py. 
With platformio and my proposed esp8266 platform, there is a bug with partition file I would have to correct.

Note that even with optimized partition, as soon as you put TCP/IP in code the code gets too big due to RTOS 3.4 new IP stack.
So if you want to stay with Sparkfun Thing there is no solution. Stay with RTOS 1.5 or buy a 5$ 16 Mbytes D1-mini on Ali Express!
